teps/0092-scheduling-timeout.md

```diff
@@ -17,361 +24,474 @@ authors:
-  * Mitigation: Consider this the last timeout related feature we add; on the next timeout related feature request
-    we consider building a separate controller to control timeouts. Not sure this is realistic though; maybe in reality
-    we're likely to end up accepting a number of different timeouts to support different scenarios.
+### Total Timeout
 
-### User Experience
+The `total` timeout of a `TaskRun` is the combination of `scheduling` and `execution` timeouts. It applies from the start time of the `TaskRun` until the completion time of the `TaskRun`. This timeout is useful when a user doesn't care to distinguish between scheduling vs execution, they only want to general timeout of a `TaskRun`.
 
-* As we add more timeout options it may become more difficult to reason about which timeout is applying when
-* Can't easily adapt this configuration to other requests (e.g. including or not including image pull time in the
-  timeout)
-* At the `PipelineRun` level, this timeout applies to individual `TaskRuns` which is not the case for any other
-  timeouts
+Defaults for `total` timeout:
+- If either `execution` or `scheduling` timeout is provided, the `total` timeout will default to `60` mins
+- Otherwise, the default `total` timeout will be calculated from `scheduling` + `execution`.
+* Note that this is different from how `pipelineRun.spec.timeouts` works: It's possible to declare PipelineRun timeouts where `timeouts.tasks` + `timeouts.finally` < `timeouts.pipeline`, and `timeouts.pipeline` always defaults to 60m and is never calculated by adding `timeouts.tasks` and `timeouts.finally`.
 
-### Performance
+### API Changes
 
-Additional timeouts to consider will mean additional `TaskRuns` queued for later reconciling within the controller.
-This would mean at most 2x the number of timeouts for each individual `TaskRun`.
+`TaskRun` has one timeout field ([docs][timeout-docs]):
 
-## Design Details
+```yaml
+spec:
+  timeout: 60s
+```
 
-This feature would be introduced [at the alpha stability level](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+To support the new timeouts we would follow, we would add a dictionary to hold configuration for `scheduling`, `execution`, and `total` timeout fields. The previous `timeout` field would be deprecated once this feature is stable.
 
-By default, when not specified, this timeout would have a value of 0, indicating that there is no scheduling timeout.
+```yaml
+spec:
+  timeouts:
+    scheduling: "5m" # optional, if specified, from TaskRun.StartTime to Pod.ScheduledTime
+    execution: "5m" # optional, if specified, from Pod.ScheduledTime to TaskRun.CompletionTime
+    total: "10m" # optional, if specified, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)
+  timeout: "10m" # existing deprecated field, from TaskRun.StartTime to TaskRun.CompletionTime (defaults to 60m)
```

Comment:
If `timeout` is different that `timeouts.total`, which one takes "priority" ? Or would it be disallowed to specify `timeout` if any `timeouts.*` is specified ?

Reply:
It would be disallowed to specify timeout if any timeouts.* is specified.